### ISSUE TITLE
Cannot detect when touch is called with belongs_to touch: true

### ISSUE THREAD

**Issue by humphreyja** [rails/rails#49324]
I can't seem to find a way to detect when a touch event is called in an `after_commit` hook when the touch event is triggered by a relational "touch: true" event. If I manually call `comment.touch` the behavior is different and I can detect that the change being made is only a touch event.

In our system we rely on `updated_at` for caching, but we also have plenty of actioncable subscriptions and various other calculations being called on `after_commit`, but we only want to actually run those events when the record was updated, not just touched from a related record. 

Is there any way to detect when a commit is caused by a `belongs_to ... touch: true`? I can't seem to find anything about it and this seems like it'd be a pretty useful thing...

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", "7.0.8"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :events, force: true do |t|
    t.string :name
    t.timestamps
  end

  create_table :posts, force: true do |t|
    t.integer :author_id
    t.string :results
    t.timestamps
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
    t.string :description
    t.timestamps
  end
end

class Event < ActiveRecord::Base
end

class Post < ActiveRecord::Base
  has_many :comments

  after_commit :create_change_event, unless: :touched_only?

  def create_change_event
    Event.create(name: 'POST')
  end

  def touched_only?
    saved_changes.keys == ['updated_at']
  end
end

class Comment < ActiveRecord::Base
  belongs_to :post, touch: true
end

class BugTest < Minitest::Test
  def test_relational_touch_should_skip_hook_but_still_touch_parents
    post = Post.create!
    comment = Comment.create!(post:)
    Event.destroy_all
    post.reload

    assert_equal 0, Event.count, 'should have no events in the database'

    comment.update(description: 'something...')
    # Post should not create a new updated event because it was touched only
    assert_equal 0, Event.count, 'should still have no events in the database as we do not want to track touch only events'
  end
end
```

### Expected behavior

`saved_changes` should include `updated_at` to prevent unnecessary operations in the `after_commit` hook.

### Actual behavior

`saved_changes` is empty, leading the system to believe that nothing was changed but the `after_commit` hook was still called.

### System configuration
**Rails version**: 7.0.8

**Ruby version**: 3.2.2

**Comment by manishElitmus** [comment]
Hi @ghiculescu and @humphreyja 

I tried to reproduce this issue

1. Here is the output of the script:

```
Fetching gem metadata from https://rubygems.org/...........
Resolving dependencies...
-- create_table(:events, {:force=>true})
D, [2023-09-21T11:45:42.525828 #50679] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "events"
D, [2023-09-21T11:45:42.526594 #50679] DEBUG -- :    (0.3ms)  CREATE TABLE "events" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "name" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
   -> 0.0086s
-- create_table(:posts, {:force=>true})
D, [2023-09-21T11:45:42.526711 #50679] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "posts"
D, [2023-09-21T11:45:42.526786 #50679] DEBUG -- :    (0.0ms)  CREATE TABLE "posts" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "author_id" integer, "results" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
   -> 0.0002s
-- create_table(:comments, {:force=>true})
D, [2023-09-21T11:45:42.527294 #50679] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "comments"
D, [2023-09-21T11:45:42.527363 #50679] DEBUG -- :    (0.0ms)  CREATE TABLE "comments" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "post_id" integer, "description" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
   -> 0.0006s
D, [2023-09-21T11:45:42.548147 #50679] DEBUG -- :    (0.1ms)  CREATE TABLE "ar_internal_metadata" ("key" varchar NOT NULL PRIMARY KEY, "value" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
D, [2023-09-21T11:45:42.554990 #50679] DEBUG -- :   ActiveRecord::InternalMetadata Load (0.1ms)  SELECT "ar_internal_metadata".* FROM "ar_internal_metadata" WHERE "ar_internal_metadata"."key" = ? LIMIT ?  [["key", "environment"], ["LIMIT", 1]]
D, [2023-09-21T11:45:42.558851 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.558973 #50679] DEBUG -- :   ActiveRecord::InternalMetadata Create (0.0ms)  INSERT INTO "ar_internal_metadata" ("key", "value", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["key", "environment"], ["value", "development"], ["created_at", "2023-09-21 06:15:42.558652"], ["updated_at", "2023-09-21 06:15:42.558652"]]
D, [2023-09-21T11:45:42.559049 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
Run options: --seed 8849

# Running:

D, [2023-09-21T11:45:42.591054 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.591213 #50679] DEBUG -- :   Post Create (0.1ms)  INSERT INTO "posts" ("created_at", "updated_at") VALUES (?, ?)  [["created_at", "2023-09-21 06:15:42.590912"], ["updated_at", "2023-09-21 06:15:42.590912"]]
D, [2023-09-21T11:45:42.591308 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.592407 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.592466 #50679] DEBUG -- :   Event Create (0.0ms)  INSERT INTO "events" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "POST"], ["created_at", "2023-09-21 06:15:42.592329"], ["updated_at", "2023-09-21 06:15:42.592329"]]
D, [2023-09-21T11:45:42.592512 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.595973 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.596049 #50679] DEBUG -- :   Comment Create (0.0ms)  INSERT INTO "comments" ("post_id", "created_at", "updated_at") VALUES (?, ?, ?)  [["post_id", 1], ["created_at", "2023-09-21 06:15:42.595574"], ["updated_at", "2023-09-21 06:15:42.595574"]]
D, [2023-09-21T11:45:42.596336 #50679] DEBUG -- :   Post Update (0.0ms)  UPDATE "posts" SET "updated_at" = ? WHERE "posts"."id" = ?  [["updated_at", "2023-09-21 06:15:42.596124"], ["id", 1]]
D, [2023-09-21T11:45:42.596375 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.596520 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.596576 #50679] DEBUG -- :   Event Create (0.0ms)  INSERT INTO "events" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "POST"], ["created_at", "2023-09-21 06:15:42.596458"], ["updated_at", "2023-09-21 06:15:42.596458"]]
D, [2023-09-21T11:45:42.596615 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.596748 #50679] DEBUG -- :   Event Load (0.0ms)  SELECT "events".* FROM "events"
D, [2023-09-21T11:45:42.596888 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.596930 #50679] DEBUG -- :   Event Destroy (0.0ms)  DELETE FROM "events" WHERE "events"."id" = ?  [["id", 1]]
D, [2023-09-21T11:45:42.596968 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.597033 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.597065 #50679] DEBUG -- :   Event Destroy (0.0ms)  DELETE FROM "events" WHERE "events"."id" = ?  [["id", 2]]
D, [2023-09-21T11:45:42.597093 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.597218 #50679] DEBUG -- :   Event Count (0.0ms)  SELECT COUNT(*) FROM "events"
D, [2023-09-21T11:45:42.597394 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.597444 #50679] DEBUG -- :   Comment Update (0.0ms)  UPDATE "comments" SET "description" = ?, "updated_at" = ? WHERE "comments"."id" = ?  [["description", "something..."], ["updated_at", "2023-09-21 06:15:42.597303"], ["id", 1]]
D, [2023-09-21T11:45:42.597597 #50679] DEBUG -- :   Post Update (0.0ms)  UPDATE "posts" SET "updated_at" = ? WHERE "posts"."id" = ?  [["updated_at", "2023-09-21 06:15:42.597492"], ["id", 1]]
D, [2023-09-21T11:45:42.597839 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.597968 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.598020 #50679] DEBUG -- :   Event Create (0.0ms)  INSERT INTO "events" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "POST"], ["created_at", "2023-09-21 06:15:42.597909"], ["updated_at", "2023-09-21 06:15:42.597909"]]
D, [2023-09-21T11:45:42.598057 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.598136 #50679] DEBUG -- :   Event Count (0.0ms)  SELECT COUNT(*) FROM "events"
F

Failure:
BugTest#test_relational_touch_should_skip_hook_but_still_touch_parents [bug_reproduce.rb:71]:
should still have no events in the database as we do not want to track touch only events.
Expected: 0
  Actual: 1


rails test bug_reproduce.rb:62



Finished in 0.009168s, 109.0750 runs/s, 218.1501 assertions/s.
1 runs, 2 assertions, 1 failures, 0 errors, 0 skips
```

If we look at the output for sql query executed

```
D, [2023-09-21T11:45:42.597394 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.597444 #50679] DEBUG -- :   Comment Update (0.0ms)  UPDATE "comments" SET "description" = ?, "updated_at" = ? WHERE "comments"."id" = ?  [["description", "something..."], ["updated_at", "2023-09-21 06:15:42.597303"], ["id", 1]]
D, [2023-09-21T11:45:42.597597 #50679] DEBUG -- :   Post Update (0.0ms)  UPDATE "posts" SET "updated_at" = ? WHERE "posts"."id" = ?  [["updated_at", "2023-09-21 06:15:42.597492"], ["id", 1]]
D, [2023-09-21T11:45:42.597839 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.597968 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.598020 #50679] DEBUG -- :   Event Create (0.0ms)  INSERT INTO "events" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "POST"], ["created_at", "2023-09-21 06:15:42.597909"], ["updated_at", "2023-09-21 06:15:42.597909"]]
```

On updating comments `description`, Post `updated_at` is getting changed in database.

But `saved_changes.keys` value is `['id', 'created_at', 'updated_at']

**Comment by maniSHarma7575** [comment]
Hi @ghiculescu and @humphreyja 

I tried to reproduce this issue

1. Here is the output of the script:

```
Fetching gem metadata from https://rubygems.org/...........
Resolving dependencies...
-- create_table(:events, {:force=>true})
D, [2023-09-21T11:45:42.525828 #50679] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "events"
D, [2023-09-21T11:45:42.526594 #50679] DEBUG -- :    (0.3ms)  CREATE TABLE "events" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "name" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
   -> 0.0086s
-- create_table(:posts, {:force=>true})
D, [2023-09-21T11:45:42.526711 #50679] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "posts"
D, [2023-09-21T11:45:42.526786 #50679] DEBUG -- :    (0.0ms)  CREATE TABLE "posts" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "author_id" integer, "results" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
   -> 0.0002s
-- create_table(:comments, {:force=>true})
D, [2023-09-21T11:45:42.527294 #50679] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "comments"
D, [2023-09-21T11:45:42.527363 #50679] DEBUG -- :    (0.0ms)  CREATE TABLE "comments" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "post_id" integer, "description" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
   -> 0.0006s
D, [2023-09-21T11:45:42.548147 #50679] DEBUG -- :    (0.1ms)  CREATE TABLE "ar_internal_metadata" ("key" varchar NOT NULL PRIMARY KEY, "value" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
D, [2023-09-21T11:45:42.554990 #50679] DEBUG -- :   ActiveRecord::InternalMetadata Load (0.1ms)  SELECT "ar_internal_metadata".* FROM "ar_internal_metadata" WHERE "ar_internal_metadata"."key" = ? LIMIT ?  [["key", "environment"], ["LIMIT", 1]]
D, [2023-09-21T11:45:42.558851 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.558973 #50679] DEBUG -- :   ActiveRecord::InternalMetadata Create (0.0ms)  INSERT INTO "ar_internal_metadata" ("key", "value", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["key", "environment"], ["value", "development"], ["created_at", "2023-09-21 06:15:42.558652"], ["updated_at", "2023-09-21 06:15:42.558652"]]
D, [2023-09-21T11:45:42.559049 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
Run options: --seed 8849

# Running:

D, [2023-09-21T11:45:42.591054 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.591213 #50679] DEBUG -- :   Post Create (0.1ms)  INSERT INTO "posts" ("created_at", "updated_at") VALUES (?, ?)  [["created_at", "2023-09-21 06:15:42.590912"], ["updated_at", "2023-09-21 06:15:42.590912"]]
D, [2023-09-21T11:45:42.591308 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.592407 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.592466 #50679] DEBUG -- :   Event Create (0.0ms)  INSERT INTO "events" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "POST"], ["created_at", "2023-09-21 06:15:42.592329"], ["updated_at", "2023-09-21 06:15:42.592329"]]
D, [2023-09-21T11:45:42.592512 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.595973 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.596049 #50679] DEBUG -- :   Comment Create (0.0ms)  INSERT INTO "comments" ("post_id", "created_at", "updated_at") VALUES (?, ?, ?)  [["post_id", 1], ["created_at", "2023-09-21 06:15:42.595574"], ["updated_at", "2023-09-21 06:15:42.595574"]]
D, [2023-09-21T11:45:42.596336 #50679] DEBUG -- :   Post Update (0.0ms)  UPDATE "posts" SET "updated_at" = ? WHERE "posts"."id" = ?  [["updated_at", "2023-09-21 06:15:42.596124"], ["id", 1]]
D, [2023-09-21T11:45:42.596375 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.596520 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.596576 #50679] DEBUG -- :   Event Create (0.0ms)  INSERT INTO "events" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "POST"], ["created_at", "2023-09-21 06:15:42.596458"], ["updated_at", "2023-09-21 06:15:42.596458"]]
D, [2023-09-21T11:45:42.596615 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.596748 #50679] DEBUG -- :   Event Load (0.0ms)  SELECT "events".* FROM "events"
D, [2023-09-21T11:45:42.596888 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.596930 #50679] DEBUG -- :   Event Destroy (0.0ms)  DELETE FROM "events" WHERE "events"."id" = ?  [["id", 1]]
D, [2023-09-21T11:45:42.596968 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.597033 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.597065 #50679] DEBUG -- :   Event Destroy (0.0ms)  DELETE FROM "events" WHERE "events"."id" = ?  [["id", 2]]
D, [2023-09-21T11:45:42.597093 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.597218 #50679] DEBUG -- :   Event Count (0.0ms)  SELECT COUNT(*) FROM "events"
D, [2023-09-21T11:45:42.597394 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.597444 #50679] DEBUG -- :   Comment Update (0.0ms)  UPDATE "comments" SET "description" = ?, "updated_at" = ? WHERE "comments"."id" = ?  [["description", "something..."], ["updated_at", "2023-09-21 06:15:42.597303"], ["id", 1]]
D, [2023-09-21T11:45:42.597597 #50679] DEBUG -- :   Post Update (0.0ms)  UPDATE "posts" SET "updated_at" = ? WHERE "posts"."id" = ?  [["updated_at", "2023-09-21 06:15:42.597492"], ["id", 1]]
D, [2023-09-21T11:45:42.597839 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.597968 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.598020 #50679] DEBUG -- :   Event Create (0.0ms)  INSERT INTO "events" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "POST"], ["created_at", "2023-09-21 06:15:42.597909"], ["updated_at", "2023-09-21 06:15:42.597909"]]
D, [2023-09-21T11:45:42.598057 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.598136 #50679] DEBUG -- :   Event Count (0.0ms)  SELECT COUNT(*) FROM "events"
F

Failure:
BugTest#test_relational_touch_should_skip_hook_but_still_touch_parents [bug_reproduce.rb:71]:
should still have no events in the database as we do not want to track touch only events.
Expected: 0
  Actual: 1


rails test bug_reproduce.rb:62



Finished in 0.009168s, 109.0750 runs/s, 218.1501 assertions/s.
1 runs, 2 assertions, 1 failures, 0 errors, 0 skips
```

If we look at the output for sql query executed

```
D, [2023-09-21T11:45:42.597394 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.597444 #50679] DEBUG -- :   Comment Update (0.0ms)  UPDATE "comments" SET "description" = ?, "updated_at" = ? WHERE "comments"."id" = ?  [["description", "something..."], ["updated_at", "2023-09-21 06:15:42.597303"], ["id", 1]]
D, [2023-09-21T11:45:42.597597 #50679] DEBUG -- :   Post Update (0.0ms)  UPDATE "posts" SET "updated_at" = ? WHERE "posts"."id" = ?  [["updated_at", "2023-09-21 06:15:42.597492"], ["id", 1]]
D, [2023-09-21T11:45:42.597839 #50679] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T11:45:42.597968 #50679] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T11:45:42.598020 #50679] DEBUG -- :   Event Create (0.0ms)  INSERT INTO "events" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "POST"], ["created_at", "2023-09-21 06:15:42.597909"], ["updated_at", "2023-09-21 06:15:42.597909"]]
```

On updating comments `description`, Post `updated_at` is getting changed in database.

But `saved_changes.keys` value is `['id', 'created_at', 'updated_at']

**Comment by alpaca-tc** [comment]
This behavior seems to be related to the change in #36219.

**Comment by humphreyja** [comment]
@maniSHarma7575 I'm not sure what `saved_changes.keys` you are looking at, I was referring to `saved_changes` on the Post object that gets touched when comments are changed. You can see in the query output that the Post `updated_at` is being changed in the database but `saved_changes.keys` is empty/does not include `updated_at` even though it was committed to the database. Because of that missing, an Event is being created.

For example, if you called `Post.touch` it would not create an event because calling `touch` directly (not through a `belongs_to ... touch: true`) does actually put `updated_at` in the `saved_changes.keys`.  It is that behavior that is missing for relational belongs_to ... touch: true that I'm addressing here.

**Comment by humphreyja** [comment]
@maniSHarma7575 If that was the case then your test would have passed.... You can see from your SQL output that the event is created which means that `saved_changes.keys` is empty. Also `id` should never be in `saved_changes.keys` unless you are creating the object or actually updating the ID, so I suspect you are looking at an old version of Post. reload that object and call an update on comments and you'll see what I am talking about.

**Comment by maniSHarma7575** [comment]
> @maniSHarma7575 If that was the case then your test would have passed.... You can see from your SQL output that the event is created which means that `saved_changes.keys` is empty. Also `id` should never be in `saved_changes.keys` unless you are creating the object or actually updating the ID, so I suspect you are looking at an old version of Post. reload that object and call an update on comments and you'll see what I am talking about.

@humphreyja 
Modified Script

```
class Post < ActiveRecord::Base
  has_many :comments

  after_commit :create_change_event, unless: :touched_only?

  def create_change_event
    #puts "I am executed after commit post created query"
    #puts saved_changes.keys
    Event.create(name: 'POST')
  end

  def touched_only?
    puts "*" * 100
    puts saved_changes.keys
    puts "*" * 100
    saved_changes.keys == ['updated_at']
  end
end
```

Output:

```
D, [2023-09-21T22:02:19.878544 #40203] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T22:02:19.878669 #40203] DEBUG -- :   Event Count (0.0ms)  SELECT COUNT(*) FROM "events"
D, [2023-09-21T22:02:19.878887 #40203] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T22:02:19.878927 #40203] DEBUG -- :   Comment Update (0.0ms)  UPDATE "comments" SET "description" = ?, "updated_at" = ? WHERE "comments"."id" = ?  [["description", "something..."], ["updated_at", "2023-09-21 16:32:19.878756"], ["id", 1]]
D, [2023-09-21T22:02:19.879094 #40203] DEBUG -- :   Post Update (0.0ms)  UPDATE "posts" SET "updated_at" = ? WHERE "posts"."id" = ?  [["updated_at", "2023-09-21 16:32:19.878973"], ["id", 1]]
D, [2023-09-21T22:02:19.879130 #40203] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
****************************************************************************************************
id
created_at
updated_at
****************************************************************************************************
D, [2023-09-21T22:02:19.879304 #40203] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-21T22:02:19.879343 #40203] DEBUG -- :   Event Create (0.0ms)  INSERT INTO "events" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "POST"], ["created_at", "2023-09-21 16:32:19.879224"], ["updated_at", "2023-09-21 16:32:19.879224"]]
D, [2023-09-21T22:02:19.879384 #40203] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-21T22:02:19.879459 #40203] DEBUG -- :   Event Count (0.0ms)  SELECT COUNT(*) FROM "events"
F

Failure:
BugTest#test_relational_touch_should_skip_hook_but_still_touch_parents [bug_reproduce.rb:76]:
should still have no events in the database as we do not want to track touch only events.
Expected: 0
  Actual: 1


bin/rails test bug_reproduce.rb:67



Finished in 0.013770s, 72.6216 runs/s, 145.2433 assertions/s.
1 runs, 2 assertions, 1 failures, 0 errors, 0 skips
```

**Comment by humphreyja** [comment]
@maniSHarma7575 Ok this is pretty interesting. Those changes are definitely coming from when the post is created. I'll update the test script to reload the post and you will see what I mean. But this does bring up an interesting point, `saved_changes` aren't reloaded at all when calling a relational touch. Those are left over from when it was created even though `after_commit` is being called again on it. That's a bit unexpected.


```ruby
class BugTest < Minitest::Test
  def test_relational_touch_should_skip_hook_but_still_touch_parents
    post = Post.create!
    comment = Comment.create!(post:)
    Event.destroy_all
    post.reload # <--- ADD THIS

    assert_equal 0, Event.count, 'should have no events in the database'

    comment.update(description: 'something...')
    # Post should not create a new updated event because it was touched only
    assert_equal 0, Event.count, 'should still have no events in the database as we do not want to track touch only events'
  end
end
```

**Comment by maniSHarma7575** [comment]
@humphreyja You are right. It's empty

**Comment by maniSHarma7575** [comment]
@ghiculescu It supposed to work this way?

**Comment by dhavalsingh** [comment]
I was facing a similar [issue](https://www.reddit.com/r/rails/comments/17h6hlj/using_touch_with_belongs_to_doesnt_resetupdate/?utm_source=share&utm_medium=web2x&context=3)

I think there are 2 things to be noted here:

> Indirect changes through touch: true: When touch: true triggers an update, it's somewhat indirect in terms of the object's loaded state within the running Ruby application. The Car object, if it was already loaded, doesn't internally register a change through the typical "dirty tracking" mechanisms because, from its perspective within the application, its attributes haven't been manually altered. The touch happens at the database level, and without reloading the Car instance in your Ruby application, you might not see this change reflected in the object's dirty tracking state.

Second is something I've also realised recently is 

> Avoid calls to update, save or other methods which create side-effects to the object inside your callback. For example, don't call update(attribute: "value") within a callback. This can alter the state of the model and may result in unexpected side effects during commit. Instead, you can safely assign values directly (for example, self.attribute = "value") in before_create / before_update or earlier callbacks.

This is given in the [docs](https://edgeguides.rubyonrails.org/active_record_callbacks.html#callback-registration).
So 
```
  def create_change_event
    Event.create(name: 'POST')
  end
```
Doing this will have some weird side effects. 

Although, i was thinking of raising a PR to update the docs a little for `touch` to mention it doesn't save the state used for dirty tracking in case of indirect touches. This should make things more clear. 

Also a problem i have with this is, indirect `touch` can trigger an after_commit but not update the dirty method tracking which is not a good idea, as a lot of folks use dirty methods to run callbacks and this can cause a lot of issues. Both of these things IMO should be mentioned somewhere

**Comment by humphreyja** [comment]
@dhavalsingh It really is the inconsistency with indirect touches triggering `after_commit` hooks while skipping all of the dirty tracking vs direct touches that do that I see as the problem. I provided a simpler case in this PR to explain the issue but what we are running into at my company is that the `Post` object kicks off an expensive calculation job. But we do not want to perform those jobs if you are only touching the object. Touching is equally important as that is used in caching. For now I've had to resort to manually touching everything with an `after_touch` callback on the `Comment`. It's a messy solution in my mind and likely to cause problems later on when we add more models and such. I also don't believe the `after_touch` callback is called with indirect callbacks either so there's another layer of fragility. 

To address the use of callbacks mentioned in the docs you linked to, that language talks about calling commitable changes to THE object, meaning the same object that is triggering the callback. That of course would cause side-effects like endless loops and such, but mutating other objects is totally fine and actually very common. (after_commit :track_event, after_update :update_user_metadata, after_destroy :start_job) The :start_job example is creating an object in redis to start a job, its a different database but it is still creating an object and some job runners are tied more into active record than others. So that case is still very much allowed.
